main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala

```diff
@@ -62,17 +62,33 @@ object Unification {
   def unifyTypes(tpe1: Type, tpe2: Type)(implicit flix: Flix): Result[Substitution, UnificationError] = {
     (tpe1, tpe2) match {
       case (x: Type.Var, y: Type.Var) =>
-        // Case 1: Check if the type variables are syntactically the same.
-        if (x.id == y.id && x.kind == y.kind)
-          return Result.Ok(Substitution.empty)
-        // Case 2: The left type variable is flexible.
-        if (x.rigidity == Rigidity.Flexible)
-          return Result.Ok(Substitution.singleton(x, y))
-        // Case 3: The right type variable is flexible.
-        if (y.rigidity == Rigidity.Flexible)
-          return Result.Ok(Substitution.singleton(y, x))
-        // Case 4: Both type variables are rigid.
-        Result.Err(UnificationError.RigidVar(x, y))
+        (x.id, y.id, x.rigidity, y.rigidity) match {
```

Comment:
I think the complexity here is getting out of hand. I wonder what can be done.

Could we not just do the kind check first and immediately return if it fails? Basically like it was in the old code?

Reply:
We can't do the kind check completely separately from rigidity check. Both kind and rigidity limit the directionality of the substitution so it's important to make sure they don't disagree. But if we start by sorting the pair by kind, that may reduce complexity.

Reply:
Presorting doesn't work because we lose equality information (e.g. if kinds are equal, either ordering should work).